### PR TITLE
User/freerb/6 create infra headers

### PR DESCRIPTION
**Pull request by BlakeFreer**
Creates the headers for infra

### CODE REVIEW

File: firmware/infra/util/noncopyable.hpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+/// @note Taken from Christopher Kormanyos
+
+#ifndef UTIL_NONCOPYABLE_HPP
+#define UTIL_NONCOPYABLE_HPP
+
+namespace util {
+
+/// @brief Privately inherit `util::noncopyable` to prevent copy constructors and
+/// reassignment.
+class noncopyable {
```
Comment (samparent97) [inline firmware/infra/util/noncopyable.hpp:12]: Let's use `UpperCamelCase` for class names

File: firmware/infra/util/noncopyable.hpp [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    noncopyable(const noncopyable &) = delete;
+    noncopyable(noncopyable &&) = delete;
+
+    auto operator=(const noncopyable &) -> noncopyable & = delete;
+    auto operator=(noncopyable &&) -> noncopyable & = delete;
+};
+
+} // namespace util
+
+#endif
```
Comment (samparent97) [inline firmware/infra/util/noncopyable.hpp:27]: Add new line

File: firmware/infra/mcal_input/input_base.hpp [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+#ifndef MCAL_INPUT_BASE_HPP
+#define MCAL_INPUT_BASE_HPP
+
+#include "util/noncopyable.hpp"
+
+namespace mcal {
+namespace input {
+
+/// @brief Base class for all microcontroller input sources
+template <typename return_type>
```
Comment (samparent97) [inline firmware/infra/mcal_input/input_base.hpp:13]: Let's use camelCase for variable names

**Review by samparent97 — COMMENTED** [review]
Just a few comments about conventions

**Review by tullom — APPROVED** [review]
LGTM

**Pull request merged**